File: ai/mcp/typescript/McpTypes.ts
import {
    ToolCallback,
    PromptCallback,
    ReadResourceCallback,
    ReadResourceTemplateCallback,
    ResourceTemplate
} from "@modelcontextprotocol/sdk/server/mcp.js";
import {
    StreamableHTTPServerTransport
} from "@modelcontextprotocol/sdk/server/streamableHttp.js";
import {
    z,
    ZodRawShape,
} from 'zod';

/**
 * Model context protocol tool.
 */
export interface McpTool {
    "name"?: string;
    "title"?: string;
    "description"?: string;
    "inputSchema"?: any;
    "parameters"?: McpToolParameters;
}

/**
 * Model context protocol tool parameters. 
 */
export interface McpToolParameters {
    "properties"?: any;
    "required"?: Array<string>;
    "type"?: string;
    "additionalProperties"?: boolean;
}

/**
 * Model context protocol tool callback.
 */
export interface McpToolCallback {
    "name"?: string;
    "callback"?: ToolCallback<undefined | ZodRawShape>;
}

/**
 * Model context protocol prompt.
 */
export interface McpPrompt {
    "name"?: string;
    "title"?: string;
    "description"?: string;
    "arguments"?: any;
}

/**
 * Model context protocol prompt callback.
 */
export interface McpPromptCallback {
    "name"?: string;
    "callback"?: PromptCallback<undefined | any /*PromptArgsRawShape*/>;
}

/**
 * Model context protocol prompt helper.
 */
export interface McpPromptHelper {
    "name"?: string;
    "prompt"?: string;
}

/**
 * Model context protocol resource.
 */
export interface McpResource {
    "name"?: string;
    "title"?: string;
    "description"?: string;
    "uri"?: string | ResourceTemplate;
    "mimeType"?: string;
}

/**
 * Model context protocol resource callback.
 */
export interface McpResourceCallback {
    "name"?: string;
    "callback"?: ReadResourceCallback | ReadResourceTemplateCallback;
}

/**
 * mcp http transport model.
 */
export interface McpHttpTransportModel {
    sessionId: string;
    transport: StreamableHTTPServerTransport;
}

/**
 * mcp function tool
 */
export interface McpFunctionTool {
    clientId: string;
    serverId: string;
    type: string;
    name: string;
    description?: string | null;
    strict: boolean;
    inputSchema?: Record<string, unknown>;
    parameters?: McpToolParameters;
}

File: ai/mcp/python/McpTypes.py
from pydantic import AnyUrl, TypeAdapter
from typing import Optional, Any, List, Union, Dict

# Model context protocol tool parameters.
class McpToolParameters:
    """
    Model context protocol tool.
    """
    def __init__(self,
                 parameters: dict[str, Any] | None = None):
        self.parameters = parameters

    def __repr__(self):
        return f"McpToolParameters(parameters={self.parameters})"

# Model context protocol tool.
class McpTool:
    """
    Model context protocol tool.
    """
    def __init__(self,
                 name: str,
                 title: str,
                 description: str,
                 inputSchema: Any,
                 parameters: McpToolParameters | None = None):
        self.name = name
        self.title = title
        self.description = description
        self.inputSchema = inputSchema
        self.parameters = parameters

    def __repr__(self):
        return f"McpTool(name={self.name}, " \
            f"title={self.title}, " \
            f"description={self.description}, " \
            f"inputSchema={self.inputSchema}, " \
            f"parameters={self.parameters})"

# Model context protocol prompt.
class McpPrompt:
    """
    Model context protocol prompt.
    """
    def __init__(self,
                 name: str,
                 title: str,
                 description: str,
                 arguments: Any):
        self.name = name
        self.title = title
        self.description = description
        self.arguments = arguments

    def __repr__(self):
        return f"McpPrompt(name={self.name}, " \
            f"title={self.title}, " \
            f"description={self.description}, " \
            f"arguments={self.arguments})"

# Model context protocol resource.
class McpResource:
    """
    Model context protocol resource.
    """
    def __init__(self,
                 name: str,
                 title: str,
                 description: str,
                 uri: AnyUrl,
                 mimeType: str):
        self.name = name
        self.title = title
        self.description = description
        self.uri = uri
        self.mimeType = mimeType

    def __repr__(self):
        return f"McpResource(name={self.name}, " \
            f"title={self.title}, " \
            f"description={self.description}, " \
            f"uri={self.uri}, " \
            f"mimeType={self.mimeType})"

# Model context protocol prompt helper.
class McpPromptHelper:
    """
    Model context protocol prompt helper.
    """
    def __init__(self,
                 name: str,
                 prompt: str):
        self.name = name
        self.prompt = prompt

    def __repr__(self):
        return f"McpPromptHelper(name={self.name}, " \
            f"prompt={self.prompt})"

# Model context protocol function tool.
class McpFunctionTool:
    """
    Model context protocol function tool.
    """
    def __init__(self,
                 clientId: str,
                 serverId: str,
                 type: str,
                 name: str,
                 description: str | None = None,
                 strict: bool | None = None,
                 inputSchema: Dict[str, object] | None = None,
                 parameters: McpToolParameters | None = None):
        self.clientId = clientId
        self.serverId = serverId
        self.type = type
        self.name = name
        self.description = description
        self.strict = strict
        self.inputSchema = inputSchema
        self.parameters = parameters

    def __repr__(self):
        return f"McpFunctionTool(name={self.name}, " \
            f"clientId={self.clientId}, " \
            f"serverId={self.serverId}, " \
            f"type={self.type}, " \
            f"description={self.description}, " \
            f"strict={self.strict}, " \
            f"inputSchema={self.inputSchema}, " \
            f"parameters={self.parameters})"